src/renderer/atlas/shader_ps.hlsl

```diff
@@ -73,18 +81,40 @@ Output main(PSData data) : SV_Target
     }
     case SHADING_TYPE_DOTTED_LINE:
     {
-        const bool on = frac(data.position.x / (2.0f * underlineWidth)) < 0.5f;
+        const bool on = frac(data.position.x / (2.0f * underlineWidth * horizontalScale)) < 0.5f;
         color = on * premultiplyColor(data.color);
         weights = color.aaaa;
         break;
     }
-    case SHADING_TYPE_DOTTED_LINE_WIDE:
+    case SHADING_TYPE_DASHED_LINE:
     {
-        const bool on = frac(data.position.x / (4.0f * underlineWidth)) < 0.5f;
+        const bool on = frac(data.position.x / (backgroundCellSize.x * horizontalScale)) < 0.5f;
         color = on * premultiplyColor(data.color);
         weights = color.aaaa;
         break;
     }
+    case SHADING_TYPE_CURLY_LINE:
+    {
+        uint cellRow = floor(data.position.y / backgroundCellSize.y);
+        // Use the previous cell when drawing 'Double Height' curly line.
+        cellRow -= verticalScale - 1;
+        const float cellTop = cellRow * backgroundCellSize.y;
+        const float centerY = cellTop + curlyLineCellOffset * verticalScale;
+        const float strokeWidthHalf = underlineWidth * verticalScale / 2.0f;
+        const float amp = curlyLinePeakHeight * verticalScale;
+        const float freq = curlyLineWaveFreq / horizontalScale;
+
+        // The wave begins with a negative peak. We phase shift the frequency by
+        // `Pi` to begin with a positive peak.
+        const float Pi = radians(180);
+        const float s = sin(data.position.x * freq + Pi);

```

Comment:
It turned out that the band-aid was applied to the wrong thing 🥲

We needed

```diff
- const float d = abs(centerY + s * amp - data.position.y);
+ const float d = abs(centerY - s * amp - data.position.y);
```

For `sin(x) == 1`, we should subtract that from the baseline, and add in case of -1.